primitives/src/toeplitz.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) 2023 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Toeplitz matrices and Circulant matrices operations
+//! References: `https://eprint.iacr.org/2020/1516.pdf`
+
+use crate::errors::PrimitivesError;
+use ark_ff::FftField;
+use ark_poly::{domain::DomainCoeff, EvaluationDomain, GeneralEvaluationDomain};
+use ark_std::{ops::Mul, string::ToString, vec::Vec};
+use jf_utils::hadamard_product;
+
+/// An `NxN` [Circulant Matrix](https://en.wikipedia.org/wiki/Circulant_matrix)
+/// is unambiguously represented by its first column, and has the form:
+///
+/// [c_0   c_N-1 c_N-2 ... c_1]
+/// [c_1   c_0   c_N-1 ... c_2]
+/// [c_2   c_1   c_0   ... c_3]
+/// [..    ..    ..    ... .. ]
+/// [c_N-1 c_N-2 c_N-3 ... c_0]
+///
+/// It is a special case of [`ToeplitzMatrix`].
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct CirculantMatrix<F: FftField> {
+    col: Vec<F>,
+}
+
+impl<F: FftField> CirculantMatrix<F> {
+    /// Construct a Circulant matrix by its first column vector
+    pub fn new(col: Vec<F>) -> Self {
+        Self { col }
+    }
+
+    /// Fast multiplication of a Circulant matrix by a vector via FFT
+    /// Details see Section 2.2.1 of [Tomescu20](https://eprint.iacr.org/2020/1516.pdf).
+    // TODO: (alex) think of ways to extend to arbitrary length vector (simply
+    // truncate doesn't work).
+    pub fn fast_vec_mul<T>(&self, m: &[T]) -> Result<Vec<T>, PrimitivesError>
+    where
+        T: for<'a> Mul<&'a F, Output = T> + DomainCoeff<F>,
+    {
+        if !m.len().is_power_of_two() {
```

Comment:
I'm not sure whether this is necessary. Such check could be performed when you initialize an evaluation domain.
IIUC ark-poly already implements mixed radix FFT so it may supports non-power-of-2.

Reply:
It's true that GeneralDomain allows for MixedRadixDomain. but here, our algorithm only works for power-of-two dim vectors.

This check plays two roles: 1) early abort and don't rely on internal logic of ark_poly; 2) avoid accidental instantiation of mixed radix domain. (which we shouldn't be using)